controlplane/kubeadm/internal/machinefilters/machine_filters_test.go

```diff
@@ -183,3 +190,154 @@ func TestInFailureDomain(t *testing.T) {
 		g.Expect(machinefilters.InFailureDomains(pointer.StringPtr("foo"), pointer.StringPtr("test"))(m)).To(BeTrue())
 	})
 }
+
+func TestMatchesKubernetesVersion(t *testing.T) {
+	t.Run("nil machine returns false", func(t *testing.T) {
+		g := NewWithT(t)
+		g.Expect(machinefilters.MatchesKubernetesVersion("some_ver")(nil)).To(BeFalse())
+	})
+
+	t.Run("nil machine.Spec.Version returns false", func(t *testing.T) {
+		g := NewWithT(t)
+		machine := &clusterv1.Machine{
+			Spec: clusterv1.MachineSpec{
+				Version: nil,
+			},
+		}
+		g.Expect(machinefilters.MatchesKubernetesVersion("some_ver")(machine)).To(BeFalse())
+	})
+
+	t.Run("machine.Spec.Version returns true if matches", func(t *testing.T) {
+		g := NewWithT(t)
+		kversion := "some_ver"
+		machine := &clusterv1.Machine{
+			Spec: clusterv1.MachineSpec{
+				Version: &kversion,
+			},
+		}
+		g.Expect(machinefilters.MatchesKubernetesVersion("some_ver")(machine)).To(BeTrue())
+	})
+
+	t.Run("machine.Spec.Version returns false if does not match", func(t *testing.T) {
+		g := NewWithT(t)
+		kversion := "some_ver_2"
+		machine := &clusterv1.Machine{
+			Spec: clusterv1.MachineSpec{
+				Version: &kversion,
+			},
+		}
+		g.Expect(machinefilters.MatchesKubernetesVersion("some_ver")(machine)).To(BeFalse())
+	})
+}
+
+func TestMatchesTemplateClonedFrom(t *testing.T) {
+	t.Run("nil machine returns false", func(t *testing.T) {
+		g := NewWithT(t)
+		g.Expect(
+			machinefilters.MatchesTemplateClonedFrom(context.TODO(), nil, nil)(nil),
+		).To(BeFalse())
+	})
+
+	t.Run("returns true if machine not found", func(t *testing.T) {
+		g := NewWithT(t)
+		client := fake.NewFakeClient()
+		kcp := &controlplanev1.KubeadmControlPlane{}
+		machine := &clusterv1.Machine{
+			Spec: clusterv1.MachineSpec{
+				InfrastructureRef: corev1.ObjectReference{
+					Kind:       "KubeadmConfig",
+					Namespace:  "default",
+					Name:       "test",
+					APIVersion: bootstrapv1.GroupVersion.String(),
+				},
+			},
+		}
+		g.Expect(
+			machinefilters.MatchesTemplateClonedFrom(context.TODO(), client, kcp)(machine),
+		).To(BeTrue())
+	})
+
+	t.Run("annotation", func(t *testing.T) {
```

Comment:
I find a little bit confusing having a table-driven test nested in a t.Run test.

Can we refactor this as flat t.Run? object shared among tests can be defined at the top of the func (I suggest to deep copy this object for each test in order to avoid side effects)

Reply:
I believe either that or removing the wrapping `t.Run()` and moving the table definition and loop to the top level would help avoid possible confusion.

Reply:
Sure. I'll refactor this. 👍

Reply:
I pulled the table-driven tests out into a separate test function specifically dealing with the CloneFrom annotations.